### ISSUE TITLE
issues when trying to create fields in __init_subclass__

### ISSUE THREAD

**Issue by quinor** [microsoft/pyright#8448]
The following code fails with `TypeVar "type[Self@Foo]" has no attribute "X"`.

```python
class Foo: # in a non-minimal example Foo is an ABC
    Y: type = type(None)

    # in a non-minimal example X may be overwritten by a subclass
    class X:
        pass

    def __init_subclass__(cls) -> None:
        class Y:
            x: cls.X = cls.X()

        cls.Y = Y
```

I'm doing a bit of metaprogramming: creating subclass attributes dynamically inside `__init_subclass__` method - but it looks like pyright has got trouble with correct recongition of `cls.X` - even though `cls` seems to get typed correctly.

**Comment by erictraut** [comment]
You are using a variable (`cls`) in a type annotation, which isn't valid. You can use a qualified class name like `Foo.X`, but you cannot use an expression with a variable in it like `cls.X`.

**Comment by erictraut** [comment]
Pyright is a static type checker. It doesn't execute of the code it's analyzing, so it's not a question of "having access to" the runtime values. Type annotations must be specified as static types. In your case, it sounds like you want to specify that instance variable `Y.x` must be an instance of `Foo.X` or a subclass thereof. The correct static type annotation that describes such a type is simply `Foo.X`.

**Comment by quinor** [comment]
I think I'm trying to implement a generic through inheritance and metaprogramming - I'm instead going to look into proper pythonic ways of implementing generics.
